File: CSharp/Tests/cbforest-sharp-tests.Shared/C4GeoTest.cs
//
// C4GeoTest.cs
//
// Author:
// 	Jim Borden  <jim.borden@couchbase.com>
//
// Copyright (c) 2016 Couchbase, Inc All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
using System;
using NUnit.Framework;
using System.IO;

namespace CBForest.Tests
{
    [TestFixture]
    public unsafe class C4GeoTest : C4Test
    {
        private static readonly string VIEW_INDEX_PATH = 
            Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "forest_temp_view.index");
        
        private C4View *_view;
        private static Random _R;
        
        public override void SetUp()
        {
            base.SetUp();
            File.Delete(VIEW_INDEX_PATH);
            C4Error error;
            _view = Native.c4view_open(_db, VIEW_INDEX_PATH, "myview", "1", C4DatabaseFlags.Create, EncryptionKey, &error);
            Assert.IsTrue(_view != null);
        }
        
        public override void TearDown()
        {
            C4Error error;
            if(_view != null) {
                Native.c4view_delete(_view, &error);   
            }
            
            base.TearDown();
        }
        
        [Test]
        public void TestCreateIndex()
        {
            CreateDocs(100);
            CreateIndex();
        }
        
        [Test]
        public void TestQuery()
        {
            const bool verbose = true;
            CreateDocs(100, verbose);
            CreateIndex();
            
            C4GeoArea area = new C4GeoArea(10, 10, 60, 60);
            C4Error error;
            C4QueryEnumerator *e = Native.c4view_geoQuery(_view, area, &error);
            Assert.IsTrue(e != null);
            
            int found = 0;
            foreach(var doc in new CBForestQueryEnumerator(e)) {
                ++found;
                var a = doc.BoundingBox;
                Assert.IsTrue(doc.Value.Equals("1234"));
                Assert.IsTrue(a.xmin <= 60 && a.xmax >= 10 && a.ymin <= 60 && a.ymax >= 10);
                Assert.IsTrue(doc.GeoJSONRaw.Equals("{\"geo\":true}"));
            }
            
            Assert.AreEqual(0, error.code);
            Assert.AreEqual(5, found);
        }
        
        private static double RandomLatitude()
        {
            return _R.NextDouble() * 180.0 - 90.0;
        }
        
        private static double RandomLongitude()
        {
            return _R.NextDouble() * 360.0 - 180.0;
        }
        
        private void CreateDocs(uint n, bool verbose = false)
        {
            _R = new Random(42);
            using(var t = new TransactionHelper(_db)) {
                for(int i = 0; i < n; i++) {
                    var docID = "doc-" + i.ToString();
                    double lat0 = RandomLatitude();
                    double lon0 = RandomLongitude();
                    double lat1 = Math.Min(lat0 + 0.5, 90.0);
                    double lon1 = Math.Min(lon0 + 0.5, 180.0);
                    var body = String.Format("({0}, {1}, {2}, {3})", lon0, lat0, lon1, lat1);
                    
                    var rq = new C4DocPutRequest();
                    rq.docID = docID;
                    rq.body = body;
                    rq.save = true;
                    C4Error error;
                    var doc = Native.c4doc_put(_db, rq, null, &error);
                    Assert.IsTrue(doc != null);
                    if(verbose) {
                        Console.WriteLine("Added {0} -> {1}", docID, body);   
                    }
                }
            }
        }
        
        private void CreateIndex()
        {
            C4Error error;
            C4Indexer* ind = Native.c4indexer_begin(_db, new C4View*[] { _view }, &error);
            Assert.IsTrue(ind != null);
            
            foreach(var doc in new CBForestDocEnumerator(ind)) {
                var body = (string)doc.GetDocument()->selectedRev.body;
                var pieces = body.Split(',');
                C4GeoArea area;
                Assert.IsTrue(Double.TryParse(pieces[0].TrimStart('('), out area.xmin));
                Assert.IsTrue(Double.TryParse(pieces[1], out area.ymin));
                Assert.IsTrue(Double.TryParse(pieces[2], out area.xmax));
                Assert.IsTrue(Double.TryParse(pieces[3].TrimEnd(')'), out area.ymax));
                var keys = new C4Key*[] { Native.c4key_newGeoJSON("{\"geo\":true}", area) };
                Assert.IsTrue(Native.c4indexer_emit(ind, doc.GetDocument(), 0, keys, new[] { "1234" }, &error));
                Native.c4key_free(keys[0]);
            }
            
            Assert.AreEqual(0, error.code);
            Assert.IsTrue(Native.c4indexer_end(ind, true, &error));
        }
    }
    
}

File: C/tests/c4GeoTest.cc
//
//  c4GeoTest.cc
//  CBForest
//
//  Created by Jens Alfke on 1/5/16.
//  Copyright © 2016 Couchbase. All rights reserved.
//

#include "c4Test.hh"
#include "c4View.h"
#include "c4DocEnumerator.h"
#include <iostream>
#ifndef _MSC_VER
#include <unistd.h>
#else
#define random() rand()
#define srandom(seed) srand(seed)
#include <algorithm>
#endif

#ifdef _MSC_VER
static const char *kViewIndexPath = "C:\\tmp\\forest_temp.view.index";
#else
static const char *kViewIndexPath = "/tmp/forest_temp.view.index";
#endif

static double randomLat() { return random() / (double)RAND_MAX * 180.0 - 90.0; }
static double randomLon() { return random() / (double)RAND_MAX * 360.0 - 180.0; }

class C4GeoTest : public C4Test {
public:

    C4View *view;

    virtual void setUp() {
        C4Test::setUp();
        ::unlink(kViewIndexPath);
        C4Error error;
        view = c4view_open(db, c4str(kViewIndexPath), c4str("myview"), c4str("1"),
                           kC4DB_Create, encryptionKey(), &error);
        Assert(view);
    }

    virtual void tearDown() {
        C4Error error;
        if (view && !c4view_delete(view, &error)) {
            fprintf(stderr, "ERROR: Failed to delete c4View: error %d/%d\n", error.domain, error.code);
            Assert(false);
        }
        c4view_free(view);
        C4Test::tearDown();
    }


    void createDocs(unsigned n, bool verbose =false) {
        srandom(42);
        TransactionHelper t(db);

        for (unsigned i = 0; i < n; ++i) {
            char docID[20];
            sprintf(docID, "%u", i);

            double lat0 = randomLat(), lon0 = randomLon();
            double lat1 = std::min(lat0 + 0.5, 90.0), lon1 = std::min(lon0 + 0.5, 180.0);
            char body[1000];
            sprintf(body, "(%g, %g, %g, %g)", lon0, lat0, lon1, lat1);

            C4DocPutRequest rq = {};
            rq.docID = c4str(docID);
            rq.body = c4str(body);
            rq.save = true;
            C4Error error;
            C4Document *doc = c4doc_put(db, &rq, NULL, &error);
            Assert(doc != NULL);
            if (verbose)
                fprintf(stderr, "Added %s --> %s\n", docID, body);
            c4doc_free(doc);
        }
    }

    void createIndex() {
        C4Error error;
        C4Indexer* ind = c4indexer_begin(db, &view, 1, &error);
        Assert(ind);

        C4DocEnumerator* e = c4indexer_enumerateDocuments(ind, &error);
        Assert(e);

        C4Document *doc;
        while (NULL != (doc = c4enum_nextDocument(e, &error))) {
            char body [1000];
            memcpy(body, doc->selectedRev.body.buf, doc->selectedRev.body.size);
            body[doc->selectedRev.body.size] = '\0';

            C4GeoArea area;
            AssertEqual(sscanf(body, "(%lf, %lf, %lf, %lf)",
                               &area.xmin, &area.ymin, &area.xmax, &area.ymax),
                        4);

            C4Key *keys[1];
            C4Slice values[1];
            keys[0] = c4key_newGeoJSON(c4str("{\"geo\":true}"), area);
            values[0] = c4str("1234");
            Assert(c4indexer_emit(ind, doc, 0, 1, keys, values, &error));
            c4key_free(keys[0]);
            c4doc_free(doc);
        }
        c4enum_free(e);
        AssertEqual(error.code, 0);
        Assert(c4indexer_end(ind, true, &error));
    }

    void testCreateIndex() {
        createDocs(100);
        createIndex();
    }

    void testQuery() {
        static const bool verbose = false;
        createDocs(100, verbose);
        createIndex();

        C4GeoArea queryArea = {10, 10, 40, 40};
        C4Error error;
        C4QueryEnumerator* e = c4view_geoQuery(view, queryArea, &error);
        Assert(e);

        unsigned found = 0;
        while (c4queryenum_next(e, &error)) {
            ++found;
            C4GeoArea a = e->geoBBox;
            if (verbose) {
                fprintf(stderr, "Found doc %.*s : (%g, %g)--(%g, %g)\n",
                    (int)e->docID.size, e->docID.buf, a.xmin, a.ymin, a.xmax, a.ymax);
            }

            C4Slice expected = C4STR("1234");
            AssertEqual(e->value, expected);
            Assert(a.xmin <= 40 && a.xmax >= 10 && a.ymin <= 40 && a.ymax >= 10);

            expected = C4STR("{\"geo\":true}");
            AssertEqual(e->geoJSON, expected);
        }
        c4queryenum_free(e);
        AssertEqual(error.code, 0);
        AssertEqual(found, 2u);
    }


    CPPUNIT_TEST_SUITE( C4GeoTest );
    CPPUNIT_TEST( testCreateIndex );
    CPPUNIT_TEST( testQuery );
    CPPUNIT_TEST_SUITE_END();
};

CPPUNIT_TEST_SUITE_REGISTRATION(C4GeoTest);
